### PR TITLE
Feature/cs4350 34 services controllers

### PR DESCRIPTION
**Pull request by Newms226**

### CODE REVIEW

File: cs4360/apps/crm/api/src/app/dao/user/user.controller.ts [hunk 1]
```diff
@@ -1,24 +1,45 @@
...
+  query: {
+    join: {
+      userType: {
+        eager: true, // deep copy down
+      },
+      contact: {
+        eager: true,
+        allow: ['id', 'firstName', 'lastName']
+      }
+    },
```
Comment (Newms226) [inline cs4360/apps/crm/api/src/app/dao/user/user.controller.ts:28]: This is how we get nested responses to a GET

File: cs4360/apps/crm/api/src/app/dao/user/user.controller.ts [hunk 2]
```diff
@@ -1,24 +1,45 @@
...
+  }
 
-@Controller('user')
-export class UserController {
-  constructor(public service: UserService) {}
+  @Override()
+  async createOne(@ParsedRequest() req: CrudRequest, @ParsedBody() dto: CreateUserRequest): Promise<CreateUserResponse> {
+    const user: UserEntity = await this.base.createOneBase(req, dto);
+    return { id: user.id };
+  }
```
Comment (Newms226) [inline cs4360/apps/crm/api/src/app/dao/user/user.controller.ts:44]: This is how we change a typical response from NestJSX/CRUD response

https://github.com/nestjsx/crud/wiki/Controllers#routes-override

File: cs4360/libs/crm/nest/dao/src/lib/entities/contact.entity.ts [hunk 3]
```diff
@@ -23,19 +24,21 @@ export class ContactEntity {
...
   @Column('varchar', {length: 250, nullable: true})
-  notes: string
+  notes?: string
 
-  @OneToMany(type => EmailEntity, email => email.contact, {
+  @ManyToMany(type => EmailEntity, email => email.contact, {
     cascade: true
   })
+  @JoinTable()
   emails?: EmailEntity[];
```
Comment (Newms226) [inline cs4360/libs/crm/nest/dao/src/lib/entities/contact.entity.ts:36]: There was no join table being created yet, now we've got one ^.^

File: cs4360/libs/crm/nest/dao/src/lib/entities/email.entity.ts [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import {Column, Entity, ManyToOne, PrimaryGeneratedColumn} from "typeorm";
-import {EmailDTO} from "@crm/crm/shared";
+import {EmailDTO} from "@crm/shared";
 import {ContactEntity} from "./contact.entity";
 
 
-@Entity('email')
+@Entity('emails')
```
Comment (Newms226) [inline cs4360/libs/crm/nest/dao/src/lib/entities/email.entity.ts:6]: semantics, but I'm anal :)

File: cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
+@Entity('users')
 export class UserEntity implements UserDTO{
 
   @PrimaryGeneratedColumn('uuid')
   id: string
 
-  @Column('varchar', { length: 50, nullable: true})
-  username: string
+  @Column('varchar', {length: 50, nullable: false, unique: true})
+  email: string
```
Comment (Newms226) [inline cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts:13]: Functional Requirement: Login via email

File: cs4360/libs/crm/shared/src/lib/contracts/user/user.create.contract.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  contact: ContactDTO;
+  email: string;
+  id: string;
+  passwordHash: string;
+  userType: UserTypeDTO;
+}
+
+export class CreateUserResponse {
+  id: string;
+}
```
Comment (Newms226) [inline cs4360/libs/crm/shared/src/lib/contracts/user/user.create.contract.ts:15]: @joliver2 @vinsonchin This is what I had in mind for the create users contract. See how these classes are used in the user controller.

These need to be created for both user & contact, for actions: create, findAll, findById, update, and delete.

File: cs4360/libs/crm/shared/src/lib/dto/contact.dto.ts [hunk 7]
```diff
@@ -1,18 +1,34 @@
...
+
+  type?: string;
+  contact: ContactDTO;
+}
+
+export interface PhoneDTO {
+  id: string;
+  number: number;
+  type?: string;
+  contact: ContactDTO;
```
Comment (Newms226) [inline cs4360/libs/crm/shared/src/lib/dto/contact.dto.ts:32]: I moved these here to try and get rid of a circular dependency, but I'm not convinced its a good idea.

@jjohnson509 @vinsonchin @GitNickProgramming @joliver2 Thoughts?

Comment (Newms226) [reply cs4360/libs/crm/shared/src/lib/dto/contact.dto.ts:32]: To be clear, it did not fix the issue

File: cs4360/tsconfig.base.json [hunk 8]
```diff
@@ -18,9 +18,7 @@
     "paths": {
       "@crm/nest/entities": ["libs/crm/nest/dao/src/lib/entities/index.ts"],
       "@crm/ng/contact-book": ["libs/crm/ng/contact-book/src/index.ts"],
-      "@crm/shared/contracts": ["libs/crm/shared/contracts/index.ts"],
-      "@crm/shared/dto": ["libs/crm/shared/dto/index.ts"],
-      "@crm/crm/shared": ["libs/crm/shared/src/index.ts"]
+      "@crm/shared": ["libs/crm/shared/src/index.ts"]
```
Comment (Newms226) [inline cs4360/tsconfig.base.json:21]: This was weird, I fixed it throughout the app

File: cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts [hunk 9]
```diff
@@ -1,19 +1,19 @@
...
 
-  @Column('varchar', {length: 50, nullable: true})
-  hash: string
+  @Column('varchar', {length: 50, nullable: true }) // TODO change to false after authn updates
+  passwordHash: string
 
   @OneToOne(type => UserTypeEntity, {cascade: true})
   @JoinColumn()
```
Comment (Newms226) [inline cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts:19]: This is the crucial piece

File: cs4360/libs/crm/nest/dao/src/lib/entities/phone.entity.ts [hunk 10]
```diff
@@ -17,5 +17,6 @@ export class PhoneEntity implements PhoneDTO{
   @ManyToOne(type => ContactEntity, contact => contact.phones, {
     onDelete: "CASCADE"
   })
+  @JoinColumn()
```
Comment (Newms226) [inline cs4360/libs/crm/nest/dao/src/lib/entities/phone.entity.ts:20]: 👍

File: cs4360/libs/crm/nest/dao/src/lib/entities/email.entity.ts [hunk 11]
```diff
@@ -18,5 +18,6 @@ export class EmailEntity implements EmailDTO{
   @ManyToOne(type => ContactEntity, contact => contact.emails, {
     onDelete: "CASCADE"
   })
+  @JoinColumn()
```
Comment (Newms226) [inline cs4360/libs/crm/nest/dao/src/lib/entities/email.entity.ts:21]: 👀

File: cs4360/apps/crm/api/src/app/dao/user/user.controller.ts (hunk 1 above)
Comment (jjohnson509) [reply cs4360/apps/crm/api/src/app/dao/user/user.controller.ts:28]: awesome, looks easy enough for future customization

File: cs4360/apps/crm/api/src/app/dao/user/user.controller.ts (hunk 2 above)
Comment (jjohnson509) [reply cs4360/apps/crm/api/src/app/dao/user/user.controller.ts:44]: good work

File: cs4360/libs/crm/nest/dao/src/lib/entities/contact.entity.ts (hunk 3 above)
Comment (jjohnson509) [reply cs4360/libs/crm/nest/dao/src/lib/entities/contact.entity.ts]: awesome!

File: cs4360/libs/crm/nest/dao/src/lib/entities/email.entity.ts (hunk 4 above)
Comment (jjohnson509) [reply cs4360/libs/crm/nest/dao/src/lib/entities/email.entity.ts:6]: lol

File: cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts (hunk 5 above)
Comment (jjohnson509) [reply cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts:13]: Should we add further validators for the entities at some point?

File: cs4360/libs/crm/shared/src/lib/dto/contact.dto.ts (hunk 7 above)
Comment (jjohnson509) [reply cs4360/libs/crm/shared/src/lib/dto/contact.dto.ts:32]: I think the `forwardRef()` from nestjs methods might help here. I tried to implement them at some point, but could never find the right place for them. My thought was that we should be able to forward the reference of the entities and that should be able to get rid of the issue.

File: cs4360/apps/crm/api/src/app/dao/user/user.controller.ts (hunk 1 above)
Comment (Newms226) [reply cs4360/apps/crm/api/src/app/dao/user/user.controller.ts:28]: 👍

File: cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts (hunk 5 above)
Comment (Newms226) [reply cs4360/libs/crm/nest/dao/src/lib/entities/user.entity.ts:13]: Yeah we totally should. See CS4360-68 Constraints for Entities and Tests to ensure them

**Pull request merged**